tests/phpunit/Environment/ChainBuildContextResolverTest.php

```diff
@@ -0,0 +1,132 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Environment;
+
+use Infection\Environment\BuildContext;
+use Infection\Environment\BuildContextResolver;
+use Infection\Environment\ChainBuildContextResolver;
+use Infection\Environment\CouldNotResolveBuildContext;
+use PHPUnit\Framework\TestCase;
+use Prophecy\Argument;
+
+final class ChainBuildContextResolverTest extends TestCase
+{
+    public function test_resolve_throws_could_not_resolve_build_context_when_chain_is_empty(): void
+    {
+        $environment = [
+            'foo' => 'bar',
+            'baz' => 'qux',
+        ];
+
+        $buildContextResolver = new ChainBuildContextResolver();
+
+        $this->expectException(CouldNotResolveBuildContext::class);
+
+        $buildContextResolver->resolve($environment);
+    }
+
+    public function test_resolve_throws_could_not_resolve_build_context_when_none_of_the_build_context_resolvers_could_resolve_the_build_context(): void
+    {
+        $environment = [
+            'foo' => 'bar',
+            'baz' => 'qux',
+        ];
+
+        $incapableBuildContextResolvers = array_map(function () use ($environment): BuildContextResolver {
+            $buildContextResolver = $this->prophesize(BuildContextResolver::class);
+
+            $buildContextResolver
+                ->resolve(Argument::is($environment))
+                ->shouldBeCalled()
+                ->willThrow(new CouldNotResolveBuildContext());
+
+            return $buildContextResolver->reveal();
+        }, range(0, 4));
+
+        $buildContextResolver = new ChainBuildContextResolver(...$incapableBuildContextResolvers);
+
+        $this->expectException(CouldNotResolveBuildContext::class);
+
+        $buildContextResolver->resolve($environment);
+    }
+
+    public function test_resolve_returns_build_context_of_first_build_context_resolver_that_could_resolve_the_build_context(): void
+    {
+        $environment = [
+            'foo' => 'bar',
+            'baz' => 'qux',
+        ];
+
+        $buildContext = new BuildContext(
+            'example/example',
+            'feature/this'
+        );
+
+        $incapableBuildContextResolvers = array_map(function () use ($environment): BuildContextResolver {
+            $buildContextResolver = $this->prophesize(BuildContextResolver::class);
+
+            $buildContextResolver
+                ->resolve(Argument::is($environment))
+                ->shouldBeCalled()
+                ->willThrow(new CouldNotResolveBuildContext());
+
+            return $buildContextResolver->reveal();
+        }, range(0, 2));
+
+        $capableBuildContextResolver = $this->prophesize(BuildContextResolver::class);
+
+        $capableBuildContextResolver
+            ->resolve(Argument::is($environment))
```

Comment:
IIRC `Argument::is()` is the default so you can just do `resolve($environment)`

Reply:
No, it's not. `Argument::exact()` is.

Reply:
👍